src/Interpreters/ExternalLoader.cpp

```diff
@@ -833,7 +851,7 @@ class ExternalLoader::LoadingDispatcher : private boost::noncopyable
                 return true; /// stop
 
             if (!min_id)
-                min_id = getMinIDToFinishLoading(forced_to_reload);
+                min_id = getMinIDToFinishLoading(forced_to_reload || force_to_update);
 
             if (info->loading_id < min_id)
                 startLoading(*info, forced_to_reload, *min_id);
```

Comment:
It [will do nothing](https://github.com/ClickHouse/ClickHouse/blob/a140f4127e67d122f18e2909387b684e959e8d3d/src/Interpreters/ExternalLoader.cpp#L905-L912) if loading is already in progress, is it expected? Also I'm afraid of deadlocks in the case when `getMinIDToFinishLoading(true)` returns `next_id_counter`, but `startLoading(*info, false, next_id_counter)` exits without updating dictionary `state_id` due to concurrent loading process with older id. It it possible that `state_id` will never become greater or equal to `next_id_counter` in this case?

Reply:
In ExternalLoader update is like reload, but using previous loadable object version. It is okay if update will not do anything if there is update or reload in progress. All bugs that exist in dictionary reload, automatically applied to update.